Data/HashSet.hs

```diff
@@ -4,25 +4,94 @@
 #endif
 
 ------------------------------------------------------------------------
--- |
--- Module      :  Data.HashSet
--- Copyright   :  2011 Bryan O'Sullivan
--- License     :  BSD-style
--- Maintainer  :  johan.tibell@gmail.com
--- Stability   :  provisional
--- Portability :  portable
---
--- A set of /hashable/ values.  A set cannot contain duplicate items.
--- A 'HashSet' makes no guarantees as to the order of its elements.
---
--- The implementation is based on /hash array mapped trie/.  A
--- 'HashSet' is often faster than other tree-based set types,
--- especially when value comparison is expensive, as in the case of
--- strings.
---
--- Many operations have a average-case complexity of /O(log n)/.  The
--- implementation uses a large base (i.e. 16) so in practice these
--- operations are constant time.
+{-|
+Module      :  Data.HashSet
+Copyright   :  2011 Bryan O'Sullivan
+License     :  BSD-style
+Maintainer  :  johan.tibell@gmail.com
+Stability   :  provisional
+Portability :  portable
+
+= Introduction
+
+'HashSet' allows you to store /unique/ elements, providing efficient insertion,
+lookups, and deletion. A 'HashSet' makes no guarantees as to the order of its
+elements.
+
+If you are storing sets of @Int@s consider using the @Data.IntSet@ from the
+<https://hackage.haskell.org/package/containers containers> package.
+
+
+== Examples
+
+All the examples below assume @HashSet@ is imported qualified, and uses the following @dataStructures@ set.
+
+>>> import qualified Data.HashSet as HashSet
+>>> let dataStructures = HashSet.fromList ["Set", "Map", "Graph", "Sequence"]
+
+=== Basic Operations
+
+Check membership in a set:
+
+>>> -- Check if "Map" and "Trie" are in the set of data structures.
+>>> HashSet.member "Map" dataStructures
+True
+>>> HashSet.member "Trie" dataStructures
+False
+
+Add a new entry to the set:
+
+>>> let moreDataStructures = HashSet.insert "Trie" dataStructures
+>>> HashSet.member \"Trie\" moreDataStructures
+> True
+
+Remove the @\"Graph\"@ entry from the set of data structures.
+
+>>> let fewerDataStructures = HashSet.delete "Graph" dataStructures
+>>> HashSet.toList fewerDataStructures
+["Map","Set","Sequence"]
+
+
+Create a new set and combine it with our original set.
+
+>>> let unorderedDataStructures = HashSet.fromList ["HashSet", "HashMap"]
+>>> HashSet.union dataStructures unorderedDataStructures
+fromList ["Map","HashSet","Graph","HashMap","Set","Sequence"]
+
+=== Using custom data with HashSet
+
+To create a @HashSet@ of your custom type, the type must have instances for 'Eq'
+and 'Hashable'. The 'Hashable' typeclass is defined in the
```

Comment:
Haddock doesn't seem to be able to link `Eq` and `Hashable` since they both aren't in scope:

```
Warning: 'Eq' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
Warning: 'Hashable' is out of scope.
    If you qualify the identifier, haddock can try to link it anyway.
```

The reason is this very limited import section below:

```
import Data.HashSet.Base
import Prelude ()
```

This would be good to fix, either by qualifying the identifiers or by modifying the imports.

Please the check the haddock warnings afterwards.

Reply:
Opted for qualifying the identifiers. Changing the imports just to reference them in Haddocks then triggers unused import warnings (since ghc can't tell that the imports are only there for Haddocks).

Reply:
Thanks!

I think you're at least in some instances linking to modules now instead of the types or type classes themselves. For example `"Data.Eq"`. I'd suggest linking to `'Data.Eq.Eq'` instead.

Reply:
Updated here and for Data.Ord.Ord.